Add questions and move the CheriBSD section (#4)
CTSRD-CHERI/cheri-faq: src/QUESTIONS.md

### What is CHERI?

> CHERI (Capability Hardware Enhanced RISC Instructions) is an international research project to revisit the fundamental security assumptions in both hardware and software. [An Introduction to CHERI](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-941.pdf) provides a high-level overview of the project. It builds upon an earlier fruit-themed project known as [BERI](https://www.cl.cam.ac.uk/research/security/ctsrd/beri/).

Reviewer: ```suggestion
[CHERI (Capability Hardware Enhanced RISC Instructions)](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/)
is a joint research project of SRI International and the University of Cambridge to revisit fundamental design choices in hardware and software to dramatically improve system security.
[An Introduction to CHERI](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-941.pdf)
provides a high-level introduction to CHERI.
```

Reviewer: I think it's fine to use the very first sentence at cheri-cpu.org for this.

Also, I'd skip the fruit bit for now. If that question appears during some conversations, I'd add it as a separate question regarding the origins of the name.

Author: That's fair.

> CHERI (Capability Hardware Enhanced RISC Instructions) is an international research project to revisit the fundamental security assumptions in both hardware and software. [An Introduction to CHERI](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-941.pdf) provides a high-level overview of the project. It builds upon an earlier fruit-themed project known as [BERI](https://www.cl.cam.ac.uk/research/security/ctsrd/beri/).

Reviewer 2: AFAIK BERI was not earlier, the two were concurrently developed, with BERI's existence being specifically to enable CHERI from the start.

> ### What types of threat has CHERI been designed to prevent?

Reviewer: ```suggestion
### What types of threats has CHERI been designed to prevent?
```

Author: 'Threat' should be singular here as it's referring to the general uncountable concept of a threat; its sense is the same as in the phrase 'threat types'.

CHERI has been developed to have two applications: memory-safe C/C++ and software compartmentalization.

> Memory-safe C/C++ includes spatial, referential, and temporal protection mechanisms, defending against both vulnerabilities and exploitation techniques. Out-of-bound writes and reads, i.e. [buffer overflows](https://en.wikipedia.org/wiki/Buffer_overflow) and over-reads, are but two examples of spatial access vulnerability that can be avoided. Exploit chains and techniques, such as the [confused deputy problem](https://en.wikipedia.org/wiki/Confused_deputy_problem), can be prevented directly.

Reviewer 2: Confused deputies still exist

> [Compartmentalizations](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/cheri-compartmentalization.html) are fine-grained models that sandbox a given application, limiting the likelihood of it crashing and resulting in a denial of service if attacked. Such a compartment therefore provides resilience to a variety of applications or services that need to demonstrate high-availability or -reliability, e.g. image processors. Where malicious payloads would ordinarily crash or compromise the image processor, the compartmentalized application would simply render the standard icon for missing images in the browser instead.

Reviewer: ```suggestion
CHERI enables fine-grained memory protection (e.g., to prevent out-of-bounds and use-after-free bugs) and highly scalable software compartmentalization (e.g., to mitigate future unknown vulnerabilities in third-party software).
[An Introduction to CHERI](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-941.pdf) describes the CHERI software model in more detail.
```

Reviewer: Note that CHERI is not specific to memory-safety issues in C/C++ but any provides features to implement memory-safety in any programming language.

Reviewer: I wouldn't want to duplicate this content as it's already described in An Introduction to CHERI. I'd also prefer to encourage a reader to read that document to understand it better.

We might want to revisit An Introduction to CHERI if it can be improved to understand the motivation better.

Author: That seems reasonable. I'll commit this suggestion, but without the reference to 'An Introduction to CHERI' as it's also duplicating content from the answer directly above.

> [Compartmentalizations](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/cheri-compartmentalization.html) are fine-grained models that sandbox a given application, limiting the likelihood of it crashing and resulting in a denial of service if attacked. Such a compartment therefore provides resilience to a variety of applications or services that need to demonstrate high-availability or -reliability, e.g. image processors. Where malicious payloads would ordinarily crash or compromise the image processor, the compartmentalized application would simply render the standard icon for missing images in the browser instead.

Reviewer 2: Compartmentalizations is a weird word to use here

   * For CHERI-RISC-V: `cheribuild.py run-riscv64-purecap -d`

> ### Where are the cheribuild.py images and builds stored?

Reviewer: What do you think about moving these two cheribuild-related questions to a separate section called Toolchains after Compartmentalization?

Author: I would keep them where they follow on from the earliest questions on cheribuild. It's usually safest to indicate to users where some output is being generated just after you've introduced the tool that actually generates it.

> ### Where are the cheribuild.py images and builds stored?

Reviewer: What do you think about moving these two cheribuild-related questions to a separate section called Toolchains after Compartmentalization?

> ### Where are the cheribuild.py images and builds stored?

Reviewer: ```suggestion
### Where are cheribuild.py images and build files stored?
```

> By default, any images will reside in `$HOME/cheri/output` and builds in `$HOME/cheri/build`.

Reviewer: ```suggestion
By default, images built by `cheribuild.py` are stored in `$HOME/cheri/output` and build files in `$HOME/cheri/build`.
```

Author: I'll include this.

> ### Can I build targets from specific branches using cheribuild.py?

Reviewer: ```suggestion
### How can I build a custom CheriBSD branch using cheribuild.py?
```

Yes. Once executed, the cheribuild.py script will create a `$HOME/cheri` directory to contain all of the repositories, builds, and configuration needed. If you change to the specific directory for your target, you will then be able to check out the correct branch and run `./cheribuild.py` against it.
```bash
cd $HOME/cheri/cheribsd
git checkout releng/22.12

cd $HOME/cheribuild
./cheribuild.py cheribsd-morello-purecap

> ```

Reviewer: ```suggestion
`cheribuild.py` implements several flags to allow building multiple CheriBSD branches and not overwrite other branches' build results:

* `--cheribsd-<target>/source-directory /path/to/your/branch` to specify a custom branch source code directory.

* `--cheribsd-<target>/build-directory /path/to/your/build` to specify where build files should be stored.

* `--cheribsd-<target>/install-directory /path/to/your/rootfs` to specify where a root filesystem files should be installed.

* `--disk-image-<target>/path /path/to/your/image.img` to specify where a disk image should be stored.
```

Author: I've a few minor changes and corrections, e.g. removing 'a' from 'where a root filesystem files'. I'd also re-word the first sentence to read: '...implements several flags to allow you to build multiple CheriBSD branches and store the results separately'. Apart from the above, I'll update it with your suggestions.

> ### Can I develop bare-metal applications for Morello?

Reviewer: ```suggestion
### How can I develop a baremetal application for Arm Morello?
```

> Yes. Arm have created an example of a [bare-metal application](https://git.morello-project.org/morello/docs/-/blob/morello/mainline/common/standalone-baremetal-readme.rst) for the board.

Reviewer: ```suggestion
See an example [baremetal application](https://git.morello-project.org/morello/docs/-/blob/morello/mainline/common/standalone-baremetal-readme.rst).
```

Author: A question phrased 'Can I do this thing?' appears neutral, at least to me, and lends itself to qualifying the response where necessary. With the phrase 'How can I do this thing?', the question presumes that the thing is possible and implies the user also has that knowledge. Partly for the sake of user experience, I would feel safer implying less background knowledge and phrasing questions to be neutral.

Reviewer: I'm fine with that. Although, I'd still change the answer not to duplicate "Arm" and "board" (as they are included in the question), and match "baremetal" with the website's contents.

> Yes. Arm have created an example of a [bare-metal application](https://git.morello-project.org/morello/docs/-/blob/morello/mainline/common/standalone-baremetal-readme.rst) for the board.

Reviewer 2: Do we care about this? This isn't our software, it isn't our hardware and we're not supporting any of it.

Author: I've no preference. It was among the participants' questions from a previous cohort and happened to be orphaned as the only question of its kind.

Reviewer: I think it would be useful to have a reference to a document describing how to run a baremetal application.

We want to include information on software stacks developed outside Cambridge/SRI in CHERI FAQ. There are people (including members of TAP) working on seL4 and FreeRTOS. There were also questions regarding baremetal applications at CHERI-CPU.

recommend you to start with CheriBSD.

### How do I get started with CheriBSD on Morello?

Follow the
[Getting Started with CheriBSD guide](https://ctsrd-cheri.github.io/cheribsd-getting-started/)
and do not skip any section.

The guide has been structured to make you aware of actions you should take
before trying to use CheriBSD.
Of critical importance, you should upgrade your Morello board firmware before
installing CheriBSD.

> 

Reviewer: If we agree on that, this should stay here.

installing CheriBSD.

### How can I upgrade CheriBSD?

Currently, there are no binary upgrades for CheriBSD.

> The only way to upgrade a CheriBSD host is to build and install CheriBSD from

Reviewer 2: I really do not want to be encouraging TAP participants to do this. The only official supported way to upgrade is to wipe and reinstall.

Author: I won't try to resolve this directly, as the affected text was part of an earlier commit, [ca1bf1d](https://github.com/CTSRD-CHERI/cheri-faq/commit/ca1bf1dcd54af840fe74dfd106b6e09bc3dcc5f9), that I was looking to reposition. I've mentioned this issue here instead: #6.

source code.
See the
[wiki page](https://github.com/CTSRD-CHERI/cheripedia/wiki/HOWTO:-Build-CheriBSD-natively-on-Morello)
for instructions how to do this.
Note that we cannot guarantee stability when upgrading CheriBSD and it is
important to make sure you can recover your data if an upgrade fails.

> ### How can I build CheriBSD from source code?

Reviewer 2: If we want to document this it should be way down at the bottom as nobody in TAP should be doing this

Alternatively, you can search for a specific package using
`pkg64c search <pattern>` and `pkg64 search <pattern>`.

> If you simply require a list of packages independently of CheriBSD, there are browsable HTML lists of the ports for both architectures: [AArch64](https://pkg.cheribsd.org/CheriBSD:20220828:aarch64.html) (hybrid) and [AArch64c](https://pkg.cheribsd.org/CheriBSD:20220828:aarch64c.html) (CHERI). In addition, there are tarballs on [pkg.cheribsd.org](https://pkg.cheribsd.org) that contain the raw JSON-formatted package data. To generate a list of CHERI packages with specific fields, you can use `curl`, `tar`, and `jq`:

Reviewer: ```suggestion
Without a CheriBSD host, you can browse CheriABI and hybrid ABI package repositories at [pkg.CheriBSD.org](https://pkg.CheriBSD.org/).

In addition, there are tarballs on [pkg.cheribsd.org](https://pkg.cheribsd.org) that contain the raw JSON-formatted package data. To generate a list of CHERI packages with specific fields, you can use `curl`, `tar`, and `jq`:
```

Reviewer: Note that `CheriBSD:20220828:aarch64{,c}.html` is not a page for the current package repositories but package repositories for this specific ABI (`CheriBSD:20220828:*`). It will change with the next release and hence the description should not be specific to a package repository for a particular ABI.

Also, the ABIs are called "the hybrid ABI" and "CheriABI" (not CHERI).

```bash
curl -O https://pkg.cheribsd.org/CheriBSD:20220828:aarch64c/packagesite.txz
tar xf *.txz
jq -j '.name," (",.version,"; ",.abi,"): ",.www,"\n"' *.yaml | sort -uf > aarch64c.txt
```
To create a similar list of hybrid packages, use the following instead:
```bash
curl -O https://pkg.cheribsd.org/CheriBSD:20220828:aarch64/packagesite.txz
```

> 

Reviewer: This seems to be redundant since we have the websites listing packages. Is there any use case for this that I'm missing?

Author: I would favour using JSON for CI, monitoring, and continuous testing over raw HTML.

Reviewer: Why would you want to monitor the list of packages? A package repository includes a stable list of packages that rarely changes. We've updated only toolchain a couple of times.

My concern is that this answer got quite lengthy while the `jq` example is not covering a case we've seen in the ecosystem. It might be better to include it in a separate Q&A once it appears.

Author: I've no objection to removing the code example and reserving it for possible future use.

Alternatively, you can search for a specific package using
`pkg64c search <pattern>` and `pkg64 search <pattern>`.

> If you simply require a list of packages independently of CheriBSD, there are tarballs on [pkg.cheribsd.org](https://pkg.cheribsd.org) that contain JSON-formatted data. To generate a list of CHERI packages with specific fields, you can use `curl`, `tar`, and `jq`:

Reviewer: Wouldn't be it better to mention the "browse" links we've added to pkg.CheriBSD.org?

Author: I'll include these as well.

[Cornucopia tutorial](https://github.com/CTSRD-CHERI/cheripedia/wiki/HOWTO:-Use-Cornucopia-with-the-22.12-CheriBSD-Release)
to read more how to use it.

> 

Reviewer: Unnecessary change.

to read more how to use it.

<!--
NOTE: When changing this title, remember to update references in this document.

> -->

Reviewer: This should be reverted.

  There are plans to make this model available for Morello as well.

> ## CheriBSD

Reviewer: I'd prefer to keep this section here because it contains more advanced questions related to CheriBSD. With the previous structure, a reader reading the questions one by one will be able to first go through more general question before they get to case-specific ones. This includes getting to know how to get started with CheriBSD, before they get to the stage they must solve CheriBSD-specific issues.

Author: With the tables of content and page separation provided by #5, I would be content reverting this move to leave the section where it was.
